### PR TITLE
Move unpacking eggs bundle to portia-entrypoint

### PR DESCRIPTION
**Pull request by vshlapakov**
The PR merges eggbased-entrypoint to portia-entrypoint and adds downloading/unpacking eggs bundle. It also changes a location of packed eggs from /app to /tmp/packed-eggs as we don't have access to modify /app from the entrypoint. Additional change is to load built-in addons eggs.

Review, please.

### CODE REVIEW

File: portia-entrypoint [hunk 1]
```diff
@@ -175,17 +175,16 @@ def main():
...
-        # Call sh_scrapy's start-crawl
-        # Replacing the process is needed so python obeys PYTHONPATH,
-        # The alternative is to modify sys.path directly but we rely
-        # on PYTHONPATH already for "scrapy list".
-        os.execve('/usr/local/bin/start-crawl', [cmd], env)
-    elif len(sys.argv) > 2:
-        os.execve(sys.argv[1], sys.argv[1:], env)
+    if len(sys.argv) > 1:
+        cmd = os.path.basename(sys.argv[1])
+        if cmd == 'start-crawl':
```
Comment (vshlapakov) [inline portia-entrypoint]: Please, note that I dropped logic for `scrapy-list` here, looks like it's useless as for now we don't use portia-stack with builder.

File: portia-entrypoint [hunk 2]
```diff
@@ -80,11 +111,54 @@ def download_portia_bundle(bundle_url, jobkey, jobauth):
...
+                os.chown(eggpath, uid, GID)
+
+
+def load_dash_eggs(eggs_path):
+    pypath = os.environ.get('PYTHONPATH', '')
+    for egg_path in glob.glob("%s/*.egg" % eggs_path):
+        if is_zipunsafe(egg_path):
+            egg_path = unpack_egg(egg_path)
+        sys.path.insert(0, egg_path)
+        pypath += '%s:' % egg_path
```
Comment (chekunkov) [inline portia-entrypoint]: if pypath is not empty and doesn't end with colon you'll get a malformed pythonpath here. also I think u may want dash eggs to be at the beginning of the path, not appended to the end

File: portia-entrypoint [hunk 3]
```diff
@@ -80,11 +111,54 @@ def download_portia_bundle(bundle_url, jobkey, jobauth):
...
+    """ Get an egg' zip unsafe flag """
+    dist = pkg_resources.EggMetadata(zipimport.zipimporter(eggpath))
+    return dist.has_metadata('not-zip-safe')
+
+
+def unpack_egg(eggpath):
+    newpath = os.path.join(DASH_UNPACKED_EGGS_PATH, os.path.basename(eggpath))
+    unpack_archive(eggpath, newpath)
+    # would be better to remove it, but we haven't permissions
+    # os.remove(eggpath)
```
Comment (chekunkov) [inline portia-entrypoint]: hm, why we don't have permission to remove from tmp?

**Review by chekunkov — CHANGES_REQUESTED** [review]

File: portia-entrypoint (hunk 3 above)
Comment (vshlapakov) [reply portia-entrypoint]: Right as now we unpack eggs bundle to tmp 👍

File: portia-entrypoint [hunk 4]
```diff
@@ -143,18 +145,26 @@ def is_zipunsafe(eggpath):
...
+    return destination
 
 
 def update_environment_for_portia():
     """Modify current environment for Portia."""
     env = os.environ.copy()
     env['TERM'] = 'xterm'
-    eggs_environment = load_dash_eggs(DASH_PACKED_EGGS_PATH)
+    # load addons_eggs in the end to give them precedence
+    eggs_environment = load_dash_eggs([DASH_PACKED_EGGS_PATH,
```
Comment (vshlapakov) [inline portia-entrypoint]: We also have to load addons eggs built-in to the stack, the only question is whether we should allow to replace the addons eggs with custom eggs or not (it's defined by precedence of the paths here). Assuming that we're moving away from eggs completely, I think it's fine to keep it as-is, but let me know if you disagree

File: portia-entrypoint [hunk 5]
```diff
@@ -97,13 +180,20 @@ def _print_flush(s):
...
     env = update_environment_for_portia()
     _print_flush("exec env: %s" % env)
-    if len(sys.argv) > 1:
-        os.execvpe(sys.argv[1], sys.argv[1:], env)
+    cmd = os.path.basename(sys.argv[0])
+    if cmd == 'start-crawl':
+        # Call sh_scrapy's start-crawl
+        # Replacing the process is needed so python obeys PYTHONPATH,
+        # The alternative is to modify sys.path directly but we rely
+        # on PYTHONPATH already for "scrapy list".
```
Comment (chekunkov) [inline portia-entrypoint]: is this note about "scrapy list" still relevant?

**Comment by vshlapakov** [comment]
@chekunkov Let me know if there's anything else awaiting for me, looks like everything is covered

**Review by chekunkov — APPROVED** [review]

**Comment by chekunkov** [comment]
@vshlapakov lgtm, do we need to wait for the new entrypoint to be released before we merge this pr, or this one is still compatible with current one?

**Comment by vshlapakov** [comment]
@chekunkov Have just checked it in a case, the PR is compatible with current entrypoint, the only drawback I should mention about is that some work is done twice, for now eggs bundle is:
- downloaded with Mesos to sandbox dir & unpacked with kumo-entrypoint to /app
- downloaded with portia-entrypoint to /tmp/ & unpacked to /tmp/unpacked-eggs

But it looks like a necessary transition state, when portia-stack is updated we're able to:
- stop providing eggs bundle with Mesos
- replace kumo-entrypoint with new version when it's ready

**Comment by chekunkov** [comment]
In this case we can merge and try it out on prod

**Pull request merged**